uencrypt: add support for mbedtls

cotequeiroz opened the pull request:
This commit includes some additional changes:
 - better handling of iv and keys in openssl/wolfssl variants
 - fix compiler warnings and whitespace
 - build all 3 variants as separate packages
 - adjust the new package name in targets' DEVICE_PACKAGES

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

---

Ping @aparcar @Ansuel @csharper2005 @system1357

I'm marking this as draft because I only briefly tested this, and more important, would like it to be tested in a target first.
If anyone test this, let me know, and I'll mark it ready to review.

cotequeiroz committed:
uencrypt: add support for mbedtls

This commit includes some additional changes:
 - better handling of iv and keys in openssl/wolfssl variants
 - fix compiler warnings and whitespace
 - build all 3 variants as separate packages
 - adjust the new package name in targets' DEVICE_PACKAGES
 - remove PKG_FLAGS:=nonshared

[Beeline SmartBox Flash - OK]
Tested-by: Mikhail Zhilkin <csharper2005@gmail.com>
[after test: replaced a hardcoded IV size of 16 by cipher_info->iv_size]
Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

aparcar commented on package/utils/uencrypt/Makefile:
```diff
@@ -4,55 +4,78 @@
 	$(INSTALL_DIR) $(1)/usr/bin
 	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/bin/uencrypt $(1)/usr/bin
 endef
 
```
```suggestion
Package/uencrypt-openssl/install = $(Package/uencrypt/default/install)
Package/uencrypt-wolfssl/install = $(Package/uencrypt/default/install)
Package/uencrypt-mbedtls/install = $(Package/uencrypt/default/install)
```

cotequeiroz replied:
Oops... 😳

aparcar reviewed (changes requested):
Missing install definition

aparcar commented:
Can you give me a trivial runtime test for this?

cotequeiroz commented:
> Can you give me a trivial runtime test for this?

This is what I used.  I got the key and iv from package/base-files/files/lib/functions/system.sh:
```
$  echo 0123456789abcde | ./uencrypt aes-128-cbc -d -n -k 2A4B303D7644395C3B2B7053553C5200 -i 00000000000000000000000000000000 | hexdump
0000000 aade 4f18 db34 2fb7 1cde 717a 2ca5 1d1a
0000010
$ echo 0123456789abcde | ./uencrypt aes-128-cbc -e -n -k 2A4B303D7644395C3B2B7053553C5200 -i 00000000000000000000000000000000 | hexdump
0000000 7ef9 79a3 93ae 26d9 cf62 2e56 79f2 41f9
0000010
$ echo 0123456789abcde | ./uencrypt aes-128-cbc -e -k 2A4B303D7644395C3B2B7053553C5200 -i 00000000000000000000000000000000 | hexdump
0000000 7ef9 79a3 93ae 26d9 cf62 2e56 79f2 41f9
0000010 cf95 b63c 7c73 548d d12a 42a4 6d3c 1c71
0000020
```

You can compare operations with what openssl would return:
```
echo 0123456789abcde | openssl aes-128-cbc -d -nopad -K 2A4B303D7644395C3B2B7053553C5200 --iv 00000000000000000000000000000000 | hexdump
0000000 aade 4f18 db34 2fb7 1cde 717a 2ca5 1d1a
0000010
```

cotequeiroz commented:
> @cotequeiroz
> 
> ```
> make[5]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
> Makefile:79: WARNING: skipping uencrypt-mbedtls -- package has no install section
> make[2]: warning: -j1 forced in submake: resetting jobserver mode.
> ```
> 
> As a result fw doesn't have uencrypt-mbedtls installed.

Sorry about that.  After applying @aparcar 's fix, it should work now.

aparcar commented:
Not sure what I'm doing wrong here but I keep getting an error about the missing `iv`

```
root@gateway:~# echo 0123456789abcde | uencrypt aes-128-cbc -e -n -k 2A4B303D7644395C3B2B7053553C5200 -i 00000000000000000000000000000000 | hexdump
Error: iv not set.
```

cotequeiroz commented:
I forgot to add `-c` before aes-128-cbc.  I suppose `getopt` has different behavior in musl compared to glibc.

cotequeiroz commented:
ECB mode is failing because of setting padding. I'll take care of that tomorrow.

csharper2005 commented:
> 
> After applying @aparcar 's fix, it should work now.

Thank you! It works. Tested on Beeline SmartBox Flash.

neheb commented:
> I forgot to add `-c` before aes-128-cbc. I suppose `getopt` has different behavior in musl compared to glibc.

yes it does

xabolcs commented on package/utils/uencrypt/src/CMakeLists.txt:
```diff
@@ -5,16 +5,27 @@ cmake_minimum_required(VERSION 2.8.12 FATAL_ERROR)
+option(USE_MBEDTLS "Use mbedTLS as crypto provider" OFF)
+if (USE_MBEDTLS)
+	if (USE_WOLFSSL)
+		message(WARNING "USE_MBEDTLS and USE_WOLFSSL are bot set. Building with USE_MBEDTLS.")
```
This warning has a typo: ` ... are both set ...`

cotequeiroz replied:
Thanks for spotting this.

csharper2005 commented:
[Beeline SmartBox Flash - OK]
Tested-by: Mikhail Zhilkin <csharper2005@gmail.com>

cotequeiroz committed:
uencrypt: split common and library-specific code

This splits the code in 4 files:
 - uencrypt.h
 - uencrypt.c - main program
 - uencrypt-openssl.c - OpenSSL/wolfSSL implementation
 - uencrypt-mbedtls - mbedTLS implementation

Other changes, accounting for ~400 bytes increase in ipk size:
 - more error condition checking and reporting,
 - hide key and iv command line arguments

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

cotequeiroz commented:
I decided to clean up the sources, to avoid duplicate code.  Since the first attempt was already tested, I added the change in a separate commit.  I tested it loca

hauke commented on package/utils/uencrypt/src/uencrypt-mbedtls.c:
```diff
@@ -0,0 +1,212 @@
+	fprintf(stderr, "Error: mbedtls_cipher_setkey: %d\n", ret);
+	goto out;
+    }
+    if (iv && (ret = mbedtls_cipher_set_iv(&ctx, iv, 16))) {
```
This assumes an ívlen of 16. I think  you should use:
````
mbedtls_cipher_set_iv(&ctx, iv, cipher_info->iv_size)
````

cotequeiroz replied:
Done.

hauke commented on package/utils/uencrypt/src/uencrypt-mbedtls.c:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    for (;;) {
+	inlen = fread(inbuf, 1, step, infile);
```
If `iv` is not set `step = mbedtls_cipher_get_block_size(&ctx);`.
If think you should make sure that inbuf is at least `mbedtls_cipher_get_block_size()` long. For most crypto algorithms  I know this is true.

cotequeiroz replied:
Good idea.  Right now the max is 16 for anything ECB.  A cipher with a large block size may later be added to the library.

I will check the value at compile-time, adjusting the buffer size if it falls under the library's maximum block length--currently 16 in mbedtls, 32 in openssl (because of XTS).

I'll make a run-time check, just in case a new version of the library introduces a cipher with block size>1024, while keeping ABI compatibility with an earlier version with a smaller value.

Should we warn--or even bail out--if that condition ever gets true?  It should fail the next operation, just as it would if it were not ECB, so I'll just set it to the buffer size and let it proceed.

Ansuel commented:
Mhhh what is the state of this? Any blocker?

Ansuel commented:
@cotequeiroz we dropped wolfssl from what i can see?

Ansuel commented:
@cotequeiroz actually wait can you force push so we can keep your gpg sign? (need to rebase as always...)

cotequeiroz commented:
@Ansuel 


> @cotequeiroz we dropped wolfssl from what i can see?

No, we did not.  I'm assuming you looked for `uencrypt-wolfss.c`, which does not exist because the WolfSSL variant uses OpenSSL API.  The only `#ifdef` blocks are to use WolfSSL headers, and the fact that WolfSSL does not print the list of supported ciphers.
